### PR TITLE
Fullscreen rewrite

### PR DESCRIPTION
**Pull request by AndreasBilke**
This PR addresses several full screen related issues.

This might solve #178 and #68 . PR #203 will be superfluous with this PR and I will close it.

@diraimondo , @svenpilz , @Lygonius and @Niklatz can you all test this PR?

### CODE REVIEW

File: src/CMakeLists.txt [hunk 1]
```diff
@@ -3,7 +3,7 @@ pkg_check_modules(GOBJECT REQUIRED gobject-2.0)
 pkg_check_modules(GIO REQUIRED gio-2.0)
 pkg_check_modules(GEE REQUIRED gee-0.8)
 pkg_check_modules(POPPLER REQUIRED poppler-glib)
-pkg_check_modules(GTK REQUIRED gtk+-3.0>=3.10)
+pkg_check_modules(GTK REQUIRED gtk+-3.0>=3.18)
```
Comment (phillipberndt) [inline src/CMakeLists.txt]: FYI, Debian stable has 3.14.5. Stretch will probably appear sometime in the next months (and has 3.22) but until then, Debian users won't be able to use the latest pdfpc anymore.

File: src/classes/window/fullscreen.vala [hunk 2]
```diff
@@ -100,19 +107,9 @@ namespace pdfpc.Window {
...
-
-                // As certain window-managers like Xfwm4 ignore movement request
-                // before the window is initially moved and set up we need to
-                // listen to this event.
-                this.size_allocate.connect(this.on_size_allocate);
-                this.configure_event.connect(this.on_configure);
-            }
-            else {
+                // start moving and fullscreening after the window was shown initially
+                this.map.connect(this.on_mapped);
```
Comment (phillipberndt) [inline src/classes/window/fullscreen.vala]: This should be `this.map_event`. See e.g. [here](https://blogs.gnome.org/jnelson/2010/10/13/those-realize-map-widget-signals/): `map` is a Gobject signal that is emitted once a map request has been sent to the X11 server, `map_event` is a signal that is emitted once the server acknowledged by sending a map event back to the client.

Comment (AndreasBilke) [inline src/classes/window/fullscreen.vala]: I fixed this

File: src/CMakeLists.txt (hunk 1 above)
Comment (AndreasBilke) [inline src/CMakeLists.txt]: fullscreen_on_monitor was added with gtk 3.18, but I added a legacy version. cmake detects old versions and use just the fullscreen api.

**Comment by diraimondo** [comment]
Under an X11 session it continues to work but, I'm sorry, it doesn't for me under a wayland one. The behavior is unchanged for me. I'm linking a live picture of both my screens (a way to make a combined screenshoot?): https://s24.postimg.org/uqt9wejxx/20170128_081745.jpg

**Comment by AndreasBilke** [comment]
I really hate GTK and its unpredictable behaviour :).

Let's see if I can get some test setup as you have.

**Comment by Niklatz** [comment]
Hey @AndreasBilke,
just build the current version of your fullscreen-rewrite branch and checked with my laptop. Everything looks great from my end. No leaking into the external monitor, no matter what screen resolution I set on the laptop monitor.
So in my opinion this pull request would fix #68!

**Comment by AndreasBilke** [comment]
@diraimondo I added one debug message in the code. could you compile and run it? I want to get a feeling about what is wrong.

**Comment by AndreasBilke** [comment]
@diraimondo I found some bug report which sounds very similar to what you have in pdfpc (https://bugs.documentfoundation.org/show_bug.cgi?id=99508). Can you also try to use `GDK_DPI_SCALE=0.5` or some other value to find out if it helps you?

**Comment by diraimondo** [comment]
@AndreasBilke this is the result with the new debug messages (same wrong scaling):
```
$ xrandr 
Screen 0: minimum 320 x 200, current 3840 x 3960, maximum 8192 x 8192
XWAYLAND2 connected 3200x1800+364+2160 290mm x 170mm
   3200x1800     59.96*+
XWAYLAND5 connected 3840x2160+0+0 600mm x 340mm
   3840x2160     59.98*+

$ bin/pdfpc ~/Desktop/pdfpc-demo.pdf                                                                                                                                                         
Screensaver inhibited
DEBUG: (0), scale 2, width 3200, height 1800, x 364, y 2160
DEBUG: (1), scale 2, width 3840, height 2160, x 0, y 0
Screensaver reactivated
```

About the suggested workaround: I can confirm that I got similar issues with presentation mode under wayland on both LibreOffice and Evince. The `GDK_DPI_SCALE=0.5` workaround works with LibreOffice but I can't see any effect with PDFPC and Evince.

**Comment by AndreasBilke** [comment]
These 3200x1800 is the native resolution, right? Very interesting, the GTK
API says that GDK Scaling is activated, but when query the monitor
resolutions it gives you the native one.

The exptected behaviour would be, that if GDK_SCALE=2 all resolutions are
diveded by 2 as well.

**Comment by diraimondo** [comment]
Yes, they are all native resolutions (as reported by xrandr).

**Comment by phillipberndt** [comment]
> The exptected behaviour would be, that if GDK_SCALE=2 all resolutions are diveded by 2 as well.

Not only is this expected, it's also documented. From the [gdk_screen_get_monitor_geometry](https://developer.gnome.org/gdk3/stable/GdkScreen.html#gdk-screen-get-monitor-geometry) API doc:

> The returned geometry is in ”application pixels”, not in ”device pixels”

Maybe this is a bug in GTK's Wayland support?

**Comment by AndreasBilke** [comment]
I found out that our set_size_request stuff makes the window too big and
gtk wayland does not scale it back when we call fullscreen.

I currently try to remove all our fixed size stuff in favor of nicely
expanding widgets.

But: diraimondo's problem seems to be a different bug than initally
described in #178 . if the other issue owners get their problem fixed with
this PR, I will merge it and we can make a new issue for diraimondo.

**Comment by diraimondo** [comment]
@AndreasBilke should we open another bug report for my problem? Today I'm testing another configuration with a low-resolution external monitor: same problem.

```
xrandr
Screen 0: minimum 320 x 200, current 4480 x 1800, maximum 8192 x 8192
XWAYLAND5 connected 3200x1800+0+0 290mm x 170mm
   3200x1800     59.96*+
XWAYLAND6 connected 1280x1024+3200+0 340mm x 270mm
   1280x1024     74.90*+
```

**Pull request merged**